database/scrapers/buildings.ts

```diff
@@ -0,0 +1,202 @@
+        `);
+    }
+    const formattedBuildingData = await Promise.all(
+      buildingsData.data.map(async (data) => {
+        const addressArray = data.addres.split(",");
+        return {
+          id: data.id,
+          identifier: data.name,
+          specialName: data.naturalName,
+          iconType: BuildingIcon.Icon,
+          addressLine1: addressArray.shift(),
+          addressLine2: addressArray.shift(),
+          latitude: data.latitude,
+          longitude: data.longitude,
+          haveFood: data.food ?? false,
+          cover: await this.uploadCoverAndGetKey(data),
+          externalDigitalGuideMode: data.externalDigitalGuideMode,
+          externalDigitalGuideIdOrURL: data.externalDigitalGuideIdOrURL,
+          createdAt: resolveDate(data.createdAt),
+          updatedAt: resolveDate(data.updatedAt),
+        };
+      }),
+    );
+    const updatedCampuses: Campus[] = [];
+    await Building.createMany(
+      formattedBuildingData.map((buildingEntry) => {
+        const campus = campusesMap.get(buildingEntry.identifier);
+        if (campus === undefined) {
+          throw new Error(`
+          No campus assigned to building ${buildingEntry.identifier} 
+          foreign key constraint will not be met for buildings table
+          `);
+        }
+        // swap campus cover placeholder (building identifier) for real cover key
+        if (
+          buildingEntry.cover !== undefined &&
+          campus.cover === buildingEntry.identifier
+        ) {
+          campus.cover = buildingEntry.cover;
+          updatedCampuses.push(campus);
+        }
+        return {
+          campusId: campus.id,
+          ...buildingEntry,
+        };
+      }),
+    );
+    task.update("Buildings created !");
+    //save changes for campuses
+    await Promise.all([...updatedCampuses].map((campus) => campus.save()));
+    task.update("campuses cover assigned !");
+  }
+
+  private async uploadCoverAndGetKey(
+    data: BuildingDraft,
+  ): Promise<string | undefined> {
+    return this.semaphore.runTask(async () => {
+      const imageKey = data.cover;
+      if (imageKey === null) {
+        this.logger.warning(`no image for building: [${data.id}]`);
+        return;
+      }
+      const extension = await this.findFileExtension(imageKey);
+      const imageStream = await this.fetchAndCheckStatus(
+        `${assetsPath}${imageKey}`,
+        `image file ${imageKey}`,
+      ).then((response) => response.body);
+      if (imageStream === null) {
+        throw new Error(
+          `No file contents for ${imageKey} for building: [${data.id}] under
+        ${assetsPath}${imageKey}`,
+        );
+      }
+      try {
+        return await this.filesService.uploadStream(
+          Readable.fromWeb(imageStream),
+          extension,
+        );
+      } catch (err) {
+        throw new Error(`failed to upload the file: ${imageKey}`, {
+          cause: err,
+        });
+      }
+    });
+  }
```

Comment:
(not an issue)
I'd usually wrap calls to the function in a semaphore.runTask() call, not the whole function definition
but I guess this works too? (time to make a semaphore decorator?)